ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditText.java

```diff
@@ -504,6 +518,22 @@ public int incrementAndGetEventCounter() {
     return ++mNativeEventCount;
   }
 
+  /**
+   * Attempt to set an accessibility error or fail silently. EventCounter is the same one used as
+   * with text.
+   *
+   * @param eventCounter
+   * @param accessibilityErrorMessage
+   */
+  public void maybeSetAccessibilityError(
+      int eventCounter, @Nullable String accessibilityErrorMessage) {
+    if (!canUpdateWithEventCount(eventCounter) || accessibilityErrorMessage == null) {
+      return;
+    }
+    announceForAccessibility("Invalid input " + accessibilityErrorMessage);
```

Comment:
What is the use case for this? As far as I know, Talkback should automatically announce error messages when they are set on the view, assuming we are firing the proper "content change" event that Talkback picks up here - https://github.com/google/talkback/blob/master/compositor/src/main/res/raw/compositor.json#L380-L387

Was this fixing a specific case?

Reply:
Thanks @blavalla. This solution had to be implemented to fix the following issue:

<details><summary>when setting accessibilityErrorAndroid onChangeText, TalkBack announces the error on the next typed character</summary>
<p>

- The user types the word `erro` in the TextInput field. 
- The user then types the last character `r` , the JavaScript component triggers the error with the prop `accessibilityErrorAndroid`

**Expected Result**:
when the user types error, the screenreader announces an error

**Actual Result**:
the screenreader does not announce the error when the text is equal to error, but on the next typed/deleted characters (errore) and (erro). There is a delay of 1 character between passing the error to the TextInput and TalkBack announcing it.

<details><summary>click to open code used in the test</summary>
<p>

```javascript
function ErrorExample(): React.Node {
  const [text, setText] = React.useState('');
  const [error, setError] = React.useState(null);
  return (
    <>
      <Button onPress={() => setError('button')} title="Press to set error" />
      <Text>
        Type error in the below TextInput to display an error message.
      </Text>
      <TextInput
        screenreaderErrorAndroid={error}
        onBlur={() => setError('onBlur')}
        onEndEditing={() => setError('onEndEditing')}
        onChangeText={newText => {
          setText(newText);
          if (newText === 'error') {
            setError('the newText is: ' + newText);
          } else if (newText === 'empty') {
            setError('');
          } else if (newText === 'null') {
            setError(null);
          } else if (error !== 'onBlur') {
            setError(null);
          }
        }}
        value={text}
        style={styles.default}
      />
    </>
  );
}
```
</p>
</details>

<details><summary>video showcasing the problem</summary>
<p>

<video src="https://user-images.githubusercontent.com/24992535/162851364-88c02366-0bbf-4174-bf5b-359c3247ab20.mp4" width="1000" />

</p>
</details>

</p>
</details>

The test case [TalkBack error announcement done on next typed character with onChangeText][11] investigates the issue in more detail:

The method maybeSetAccessibilityError is called from [fabric state updates][1] after the TalkBack accessibility announcement ([screenshot 1][3], [screenshot 2][4], [screenshot 3][5]).
The solution for this issue is triggering manually the accessibility announcement from `maybeSetAccessibilityError` and clearing  R.id.accessibility_error before the user types the next letter.

- `announceForAccessibility` triggers the accessibility error onChangeText
- `setTag(R.id.accessibility_error, null)`  clear the error before the next character is typed, otherwise the error is announced also on the next/previous character

```java
public void maybeSetAccessibilityError(
    int eventCounter, @Nullable String accessibilityErrorMessage) {
  if (!canUpdateWithEventCount(eventCounter) || accessibilityErrorMessage == null) {
    return;
  }
  announceForAccessibility("Invalid input " + accessibilityErrorMessage);
  setTag(R.id.accessibility_error, null);
}
```

[1]: https://github.com/fabriziobertoglio1987/react-native/blob/994089cfbe106ac839a07b90cb4e851a3f0aa7e3/ReactCommon/react/renderer/components/textinput/androidtextinput/react/renderer/components/androidtextinput/AndroidTextInputShadowNode.cpp#L118 "fabric updateStateIfNeeded" 
[2]: https://user-images.githubusercontent.com/24992535/162907581-c03239e7-233a-4e24-b4e6-fbf493a69040.png "state update"
[3]: https://user-images.githubusercontent.com/24992535/162907581-c03239e7-233a-4e24-b4e6-fbf493a69040.png "screenshot -  the state sent from fabric"
[4]: https://user-images.githubusercontent.com/24992535/162907091-f2e779fe-91ea-45c8-b83f-ccc50af0a361.png "screenshot - the ScreenReader announcement is done before receiving state update from fabric"
[5]: https://user-images.githubusercontent.com/24992535/162903238-48a7eb45-74ff-460a-b9a8-29256347a2b1.png "screenshot - after receiving the state update from fabric"

[10]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1095709282 "when setting error onChangeText, announcement is done on the next typed character"
[11]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1096086753 "TalkBack error announcement done on next typed character with onChange"

Reply:
Thanks, @blavalla. The new solution I implemented is:

1) The method `maybeSetAccessibilityError` receives updates from fabric (c++).
     - The method is called when the prop screenreaderErrorAndroid is updated in the TextInput onChangeText callback
     - The method triggers the event `TYPE_WINDOW_CONTENT_CHANGED`

https://github.com/facebook/react-native/blob/0513007362b419c8762a0ba9b954c231d6cb595b/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditText.java#L546-L555

2) `requestSendAccessibilityEvent` is invoked to fix the Issue [TalkBack error announcement done on next typed character with onChangeText][1], avoiding to announce the error message twice (on text "error" and next character "errore").

[1]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1096086753

https://github.com/facebook/react-native/blob/0513007362b419c8762a0ba9b954c231d6cb595b/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditText.java#L176-L181

The solution is taken from the [TextView#setError][2] method which calls `notifyViewAccessibilityStateChangedIfNeeded`. I remain available to do further improvements. :pray: 

[2]: https://github.com/aosp-mirror/platform_frameworks_base/blob/1ac46f932ef88a8f96d652580d8105e361ffc842/core/java/android/widget/TextView.java#L7284-L7285

Reply:
Thanks @blavalla, I updated the videos of the examples below.

## Test Plan

**PR Branch**
[1]. Test Cases of the functionality (Fabric) ([link][1])
[2]. Test Cases of the functionality (Paper) ([link][2])

**Issues Solved**
[6]. TalkBack error does not clear error on the next typed character when using onChangeText ([link][6])

<!-- Test Cases -->
[1]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1072101477 "Test Cases of the functionality (Fabric)"
[2]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1105964322 "Test Cases of the functionality (Paper)"
[4]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1087020844 "Runtime Error in main branch when passing value of 1 to TextInput  placeholder prop"
[5]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1095709282 "when setting accessibilityErrorAndroid onChangeText, TalkBack announces the error on the next typed character"
[6]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1096086753 "TalkBack error announcement done on next typed character with onChangeText"
[7]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1082594363 "setting the TextInput errorMessage state with setTextAndSelection Java API from JavaScript"
[8]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1082598745 "Setting the TextInput errorMessage state from fabric TextInput internal state to Java ReactTextUpdate API"